### PR TITLE
Talker CUDA graph for Qwen-3 TTS

### PR DESCRIPTION
**Pull request by evezhier**
## Purpose
This PR enables CUDA graph for Qwen3 TTS with batching support.

## Test Plan
[Done] Performance benchmark
[Done] Numerical correctness tests
[Done] Batching support
[In progress] Accuracy issue with increased concurrencies
[TODO] Benchmark update for batching

### Performance (L40S)
`vllm serve Qwen/Qwen3-TTS-12Hz-0.6B-CustomVoice --stage-configs-path vllm_omni/model_executor/stage_configs/qwen3_tts.yaml --omni --port 8000 --trust-remote-code`

`vllm-omni bench serve --backend openai-audio-speech --endpoint /v1/audio/speech --model Qwen/Qwen3-TTS-12Hz-0.6B-CustomVoice --dataset-name hf --dataset-path philschmid/mt-bench --skip-chat-template --num-prompts 10 --max_concurrency $MAX_CONCURRENCY --percentile-metrics e2el,audio_rtf,audio_duration --extra-body '{"voice": "Vivian", "instructions": "Speak with great enthusiasm", "max_new_tokens": 1024}'`

#### Main
```
============ Serving Benchmark Result ============                                                                                                                                        
Successful requests:                     10                                                                                                                                               
Failed requests:                         0                                                                                                                                                
Maximum request concurrency:             1                                                                                                                                                
Benchmark duration (s):                  117.77                                                                                                                                           
Request throughput (req/s):              0.08                                                                                                                                             
Peak concurrent requests:                2.00                                                                                                                                             
----------------End-to-end Latency----------------                                                                                                                                        
Mean E2EL (ms):                          11776.92                                                                                                                                         
Median E2EL (ms):                        7764.31                                                                                                                                          
P99 E2EL (ms):                           32342.57                                                                                                                                         
================== Text Result ===================                                                                                                                                        
Total input tokens:                      785                                                                                                                                              
Total generated tokens:                  0                                                                                                                                                
Output token throughput (tok/s):         0.00      
Peak output token throughput (tok/s):    1.00      
Peak concurrent requests:                2.00      
Total Token throughput (tok/s):          6.67       
================== Audio Result ==================  
Total audio duration generated(s):       238.16     
Total audio frames generated:            5715840    
Audio throughput(audio duration/s):      2.02       
-----------------Real Time Factor-----------------  
Mean AUDIO_RTF:                          0.48      
Median AUDIO_RTF:                        0.47       
P99 AUDIO_RTF:                           0.58      
------------------Audio Duration------------------                                                                                                                                        
Mean AUDIO_DURATION (s):                 23.82                                                                                                                                            
Median AUDIO_DURATION (s):               16.72                                                                                                                                            
P99 AUDIO_DURATION (s):                  60.49      
==================================================


============ Serving Benchmark Result ============
Successful requests:                     10        
Failed requests:                         0         
Maximum request concurrency:             8         
Benchmark duration (s):                  59.90     
Request throughput (req/s):              0.17      
Peak concurrent requests:                9.00      
----------------End-to-end Latency----------------
Mean E2EL (ms):                          29868.76  
Median E2EL (ms):                        26344.04  
P99 E2EL (ms):                           59847.30  
================== Text Result ===================
Total input tokens:                      785       
Total generated tokens:                  0         
Output token throughput (tok/s):         0.00      
Peak output token throughput (tok/s):    8.00      
Peak concurrent requests:                9.00      
Total Token throughput (tok/s):          13.11     
================== Audio Result ==================
Total audio duration generated(s):       237.60    
Total audio frames generated:            5702400   
Audio throughput(audio duration/s):      3.97      
-----------------Real Time Factor-----------------
Mean AUDIO_RTF:                          1.40      
Median AUDIO_RTF:                        1.35      
P99 AUDIO_RTF:                           2.34      
------------------Audio Duration------------------
Mean AUDIO_DURATION (s):                 23.76     
Median AUDIO_DURATION (s):               18.48     
P99 AUDIO_DURATION (s):                  55.49     
==================================================
```
#### This PR
```
============ Serving Benchmark Result ============
Successful requests:                     10        
Failed requests:                         0         
Maximum request concurrency:             1         
Benchmark duration (s):                  40.97     
Request throughput (req/s):              0.24      
Peak concurrent requests:                2.00      
----------------End-to-end Latency----------------
Mean E2EL (ms):                          4096.87   
Median E2EL (ms):                        2811.94   
P99 E2EL (ms):                           10299.14  
================== Text Result ===================
Total input tokens:                      785       
Total generated tokens:                  0         
Output token throughput (tok/s):         0.00      
Peak output token throughput (tok/s):    1.00      
Peak concurrent requests:                2.00      
Total Token throughput (tok/s):          19.16     
================== Audio Result ==================
Total audio duration generated(s):       238.16    
Total audio frames generated:            5715840   
Audio throughput(audio duration/s):      5.81      
-----------------Real Time Factor-----------------
Mean AUDIO_RTF:                          0.17      
Median AUDIO_RTF:                        0.17      
P99 AUDIO_RTF:                           0.18      
------------------Audio Duration------------------
Mean AUDIO_DURATION (s):                 23.82     
Median AUDIO_DURATION (s):               16.72     
P99 AUDIO_DURATION (s):                  60.49     
==================================================
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_talker.py [hunk 1]
```diff
@@ -1557,15 +1566,45 @@ def _talker_and_collect_speaker(ws: Iterable[tuple[str, torch.Tensor]]):
...
 
-    @torch.inference_mode()
     def talker_mtp(
         self,
         input_ids: torch.Tensor,
         input_embeds: torch.Tensor,
         last_talker_hidden: torch.Tensor,
         text_step: torch.Tensor,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        with record_function_or_nullcontext("talker_mtp"):
```
Comment (nekorobov) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_talker.py:1605]: Add docstring to the function, please

File: vllm_omni/model_executor/models/qwen3_tts/cuda_graph_talker_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
+import torch
```
Comment (nekorobov) [inline vllm_omni/model_executor/models/qwen3_tts/cuda_graph_talker_wrapper.py:1]: Add docstrings to the class and key functions, please. Specify the assumptions and the API for the GraphWrapper

File: vllm_omni/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -84,8 +84,15 @@ def load_model(self, *args, **kwargs) -> None:
             # have a separate .talker sub-module.  TTS models' code predictor
             # has internal AR loops / torch.multinomial — not graph-safe.
             has_separate_talker = getattr(self.model, "talker", None) is not None
-            if cudagraph_mode.has_full_cudagraphs() and has_separate_talker:
-                self.talker_mtp = CUDAGraphWrapper(talker_mtp, self.vllm_config, runtime_mode=CUDAGraphMode.FULL)
+            if cudagraph_mode.has_full_cudagraphs():
+                if has_separate_talker:
```
Comment (nekorobov) [inline vllm_omni/worker/gpu_model_runner.py:110]: What is the difference between these two approaches?

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_talker.py
Comment (nekorobov) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_talker.py:1]: > Limitation: BS=1, serving should be run with async scheduling disabled, but the code can be further extended to support batching and async scheduling.

I am not sure what is the best place to check these conditions. Probably, somewhere around this file. But this must be checked programmatically.

Comment (DomBrown) [reply vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_talker.py:1]: Maybe we could check the shape of the batch dimension [here](https://github.com/vllm-project/vllm-omni/pull/1925/changes#diff-17d517d4e3b0e0fe7a78bb081e76d2dc0b316c30e57294cee0afb3990b513ec2R129) before replay?

Trigger some assertion if > 1

File: vllm_omni/worker/gpu_model_runner.py (hunk 3 above)
Comment (evezhier) [reply vllm_omni/worker/gpu_model_runner.py:110]: Custom (model-specific) cuda graph wrappers instead of vLLM's default CUDAGraphWrapper. See above: 

> Only wrap talker_mtp in CUDAGraphWrapper for Omni models that have a separate .talker sub-module

File: vllm_omni/model_executor/models/qwen3_tts/cuda_graph_talker_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            layer0_code=self.input_ids_buf,
+            layer0_embed=self.last_id_hidden_buf,
+            last_talker_hidden=self.past_hidden_buf,
+            do_sample=True,
+            temperature=self.temperature,
+            top_k=self.top_k,
+        )
+        self.audio_codes_buf.copy_(audio_codes)
+
+        residual_ids = audio_codes[:, 1:]
```
[the remark this answers is not in the document]
Comment (evezhier) [reply vllm_omni/model_executor/models/qwen3_tts/cuda_graph_talker_wrapper.py]: fixed

**Comment by hsliuustc0106** [comment]
why audio results are the same?

------------------Audio Duration------------------                                           
Mean AUDIO_DURATION (s):                 23.82                                               
Median AUDIO_DURATION (s):               16.72                                               
P99 AUDIO_DURATION (s):                  60.49

**Comment by evezhier** [comment]
> why audio results are the same?
> 
> ------------------Audio Duration------------------ Mean AUDIO_DURATION (s): 23.82 Median AUDIO_DURATION (s): 16.72 P99 AUDIO_DURATION (s): 60.49

@hsliuustc0106 because I'm in the middle of updating the benchmarks after the rebase and batching support update :) I'll mark this draft for now, but the code is here, the numbers are arriving shortly.

**Comment by evezhier** [comment]
Closing this as incompatible with #1913
